test/functional/runner/index.js

```diff
@@ -0,0 +1,590 @@
+        opOptions.returnOriginal = operation.arguments[key] === 'Before' ? true : false;
+        return;
+      }
+
+      if (key === 'options') {
+        Object.assign(opOptions, operation.arguments[key]);
+        if (opOptions.readPreference) {
+          opOptions.readPreference = normalizeReadPreference(opOptions.readPreference.mode);
+        }
+
+        return;
+      }
+
+      if (key === 'readPreference') {
+        opOptions[key] = normalizeReadPreference(operation.arguments[key].mode);
+        return;
+      }
+
+      opOptions[key] = operation.arguments[key];
+    });
+  }
+
+  if (
+    args.length === 0 &&
+    !isTransactionCommand(operationName) &&
+    !isTestRunnerCommand(operationName)
+  ) {
+    args.push({});
+  }
+
+  if (Object.keys(opOptions).length > 0) {
+    // NOTE: this is awful, but in order to provide options for some methods we need to add empty
+    //       query objects.
+    if (operationName === 'distinct') {
+      args.push({});
+    }
+
+    args.push(opOptions);
+  }
+
+  let opPromise;
+  if (operationName === 'find' || operationName === 'aggregate') {
+    // `find` creates a cursor, so we need to call `toArray` on it
+    const cursor = obj[operationName].apply(obj, args);
+    opPromise = cursor.toArray();
+  } else {
+    // wrap this in a `Promise.try` because some operations might throw
+    opPromise = Promise.try(() => obj[operationName].apply(obj, args));
+  }
+
+  if (operation.result) {
+    const result = operation.result;
+
+    if (
+      result.errorContains ||
+      result.errorCodeName ||
+      result.errorLabelsContain ||
+      result.errorLabelsOmit
+    ) {
+      return opPromise
+        .then(() => {
+          throw new Error('expected an error!');
+        })
+        .catch(err => {
+          const errorContains = result.errorContains;
+          const errorCodeName = result.errorCodeName;
+          const errorLabelsContain = result.errorLabelsContain;
+          const errorLabelsOmit = result.errorLabelsOmit;
+
+          if (errorLabelsContain) {
+            expect(err).to.have.property('errorLabels');
+            expect(err.errorLabels).to.include.members(errorLabelsContain);
+          }
+
+          if (errorLabelsOmit) {
+            if (err.errorLabels && Array.isArray(err.errorLabels) && err.errorLabels.length !== 0) {
+              expect(errorLabelsOmit).to.not.include.members(err.errorLabels);
+            }
+          }
+
+          if (operation.result.errorContains) {
+            expect(err.message).to.match(new RegExp(escape(errorContains), 'i'));
+          }
+
+          if (errorCodeName) {
+            expect(err.codeName).to.equal(errorCodeName);
+          }
+
+          if (!options.swallowOperationErrors) {
+            throw err;
+          }
+        });
+    }
+
+    return opPromise.then(opResult => {
+      const actual = extractCrudResult(opResult, operation);
+      expect(actual).to.matchTransactionSpec(operation.result);
```

Comment:
If we are going to use this for more than the transaction spec test runner, we should rename it to something like `matchMongoSpec`

Reply:
:+1: